man/man1/pmlogger_merge.1
'\"macro stdmacro
.\"
.\" Copyright (c) 2013-2019 Red Hat.
.\" Copyright (c) 2000 Silicon Graphics, Inc.  All Rights Reserved.
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.TH PMLOGGER_MERGE 1 "PCP" "Performance Co-Pilot"
.SH NAME
\f3pmlogger_merge\f1 \- helper script to merge Performance Co-Pilot archives
.SH SYNOPSIS
.B $PCP_BINADM_DIR/pmlogger_merge
[\f3\-fNVE?\f1]
[\f2input-basename\f1 ... \f2output-name\f1]
.SH DESCRIPTION
.B pmlogger_merge
is a helper script that is used by
.BR pmlogger_daily (1)
when merging all of the Performance Co-Pilot (PCP)
archives for a single host on a single day
into a combined PCP archive.
.PP
.B pmlogger_merge
is a wrapper script for
.BR pmlogextract (1)
that merges all of the archives matching the
.I input-basename
arguments and creates a new archive using
.I output-name
as the base name for the physical files that constitute
an archive.
The
.I input-basename
arguments may contain meta characters in the style of
.BR sh (1).
.PP
If any of the files that are components of the
.I input-basename
archives has been compressed then on success
.B pmlogger_merge
will compress
.I output-name
using
.BR pmlogcompress (1).
.SH OPTIONS
The available command line options are:
.TP 5
\fB\-f\fR, \fB\-\-force\fR
This option causes all of the input files to be removed once the output
archive has been created.
.TP
\fB\-N\fR, \fB\-\-showme\fR
Runs
.B pmlogger_merge
in a ``show me'' or ``dry run'' mode where the
tasks that would be performed are reported, but no changes are made.
This is typically used for debugging in combination with one (verbose)
or two (very verbose)
.B \-V
options.
.TP
\fB\-V\fR, \fB\-\-verbose\fR
Increase verbosity in the output progress report.
.TP
\fB\-E\fR, \fB\-\-expunge\fR
When metric metadata differences are encountered between archives to be merged,
the default behaviour is for
.B pmlogger_merge
to fail with a fatal error and the \f2output-name\f1 archive is not created.
The
.B \-E
flag instead issues a warning and excludes the affected metrics from the merged
output archive.
This is intended for automated daily archive rotation where it is highly desirable
for unattended daily archive merging, rewriting and compression to succeed.
See also
.BR pmlogger_daily (1)
and the
.B \-x
flag to
.BR pmlogextract (1).
.TP
\fB\-?\fR, \fB\-\-help\fR
Display usage message and exit.
.SH PCP ENVIRONMENT
Environment variables with the prefix \fBPCP_\fP are used to parameterize
the file and directory names used by PCP.
On each installation, the
file \fI/etc/pcp.conf\fP contains the local values for these variables.
The \fB$PCP_CONF\fP variable may be used to specify an alternative
configuration file, as described in \fBpcp.conf\fP(5).
.SH SEE ALSO
.BR PCPIntro (1),
.BR pmlogcompress (1),
.BR pmlogextract (1),
.BR pmlogger (1)
and
.BR pmlogger_daily (1).


CHANGELOG
pcp-6.3.6 (17 March 2025)
    - Debian build changes - see debian/changelog for details

pcp-6.3.4 (13 March 2025)
    - libpcp: fix endian issue in big-endian __pmTimestamp writing
    - pcp-xsos: fine-tune error handling on bad command line options
    - pcp-xsos: single-line pmPrintValue/pminfo values, escape shell chars
    - pcp-atop: correct BPF per-process network metric names
    - pmcd, pmdaproc: allow non-local username connection attributes
    - pmlogger, libpcp: fix handling of dynamic instance domain labels
    - pmrep: use BPF instead of BCC for proc networking metrics
    - pmlogconf: use libbpf over BCC for networking metrics
    - pmlogextract: fix logic around instances labels, allowing set updates
    - pmdalinux: mem.numa.hugepages metrics (per node, per hugepage size)
    - pmdabpftrace: fix for some python3 versions async IO create_task
    - pmdanvidia: if present use dso from the libnvidia-ml runtime package
    - pmdadm, pmieconf: default to enabling the dmthin pmie rules
    - configure.ac: fix autoreconf, add AC_DEFINE_UNQUOTED descriptions
    - rpm builds: enable bpf, bpftrace, bcc on riscv64 builds
    - rpm builds: drop user creation scriptlets for rpm builds
    - deb build: force debian/rules to be recreated each time
    - containers: updates to the archive-analysis container
    - selinux: several policy updates for new AVC denials
    - docs: improve SSL/TLS diagnostic in pcp-verify script

pcp-6.3.3 (29 January 2025)
    - Debian build changes - see debian/changelog for details

pcp-6.3.3 (29 January 2025)
    - Debian build changes - see debian/changelog for details

pcp-6.3.2 (6 November 2024)
    - PMDA additions, enhancements and bug fixes:
	pmdabpf: avoid a too-low rlimit setting within pmdabpf
	pmdaproc: add an access control mechanism for remote sampling
	pmdalinux: ensure divide-by-zero cannot happen with tmpfs.full
	pmdalinux: additional MPTcpExt netstat metrics
	pmdastatsd: keep correct track of pmns mapped metric count
	pmdaopenmetrics: run as pcp user rather than root by default
	pmdaopenmetrics: ensure correct cache file permissions used
    - Server-side utilities and log management scripts:
	libpcp, pmlogger: add support for zstd(1) compression of archives
	pmlogbasename: new, maps filename to PCP archive basename
	pmlogcompress: new, archive compression/decompression utility
	pmlogcheck: performance improvements
	pmlogcheck:  allow multiple archive options on command line
	pmlogextract: extend -v option with the full pmlogger syntax
	pmlogextract: fix for reporting file size on 32-bit platforms
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: fix 64-bit big endian v3 archive timestamp decoding
	libpcp: change in archive value interpolation for non-counter metrics
	python api: allow timeval and timespec init with float/datetime
    - Client tools and utilities:
	pminfo: minor error path reworking
	pcp-buddyinfo: addition of a --no-interpol option
	pcp-ps: added option to view full task name with its argument
    - Build, infrastructure and packaging updates:
	build: deal with botched deb package names from Debian
	build: ensure pmie config.default created with correct ownership
	systemd: move pcp-reboot-init rpm spec logic from zeroconf to pcp
	selinux: ensure farm pmlogger/pmie processes labelled correctly
    - Documentation:
	docs: refine pmdaproc -A description details
	docs: make pmstat(1) match the code for -t defaults

pcp-6.3.1 (17 September 2024)
    - Client tools and utilities:
	pcp2arrow: fix extraction of values for metrics with indoms
	pcp-htop: fix platform setting enabling screen tabs
	pcp-htop: update to latest upstream bug fixes
	pcp-meminfo: added timestamp string for consistency
    - PMDA additions, enhancements and bug fixes:
	pmdaopenmetrics: add vLLM metrics support by default
	pmdahacluster: handle pacemaker 2.1.6 crm_mon output changes
	pmdahacluster: handle new style Pacemaker role descriptions
	pmdalinux: MPJoinSynBackupRx, MPJoinSynAckBackupRx netstat metrics
	pmdalinux: sysfs fallback from hinv.cpu.clock metric
	pmdaproc: lazy initialisation for accounting metrics
	pmdapmcd: expose pmcd limits as new metrics
	pmdammv: use isnan instead of open-coding NaN checks
    - Server-side utilities and log management scripts:
	pmcd: put an explicit cap on max number of fetched PMIDs
	pmcd: put an explicit cap on max number of contexts per client
	pmcd: cleanup caching of local hostname string
	pmpost: guard against possible symlink attack
	pmproxy: transition to dual Redis/Valkey support
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp_web: fix a race in webgroup context teardown on timeout
	libpcp_web: transition to dual Redis/Valkey support
	libpcp: auto-load the proc PMDA in local context mode
	libpcp: fix buffer overflow in __pmServerAddInterface
	libpcp: refactor the attr PDU content checking
	libpcp: desc PDU consistency check and error message rework
	libpcp: text PDU consistency check and error message rework
	libpcp: creds PDU consistency check and error message rework
	libpcp: label PDU consistency check and error message rework
	libpcp: instance PDU consistency check and error message rework
	libpcp: rework some fetch PDU integrity checks and error messages
	libpcp: reworked __pmDecodeIDList integrity checks and warnings
	libpcp: ensure no fd leakage to child from __pmProcessPipe
	libpcp: ensure container names contain limited characters
	libpcp: fix buffer overrun test in __pmDecodeValueSet
	libpcp: hardening of the results PDU handling
	libpcp: hardening of the creds PDU handling
	libpcp: add new __pmCtlDebug() function
	python API: fix MMV metric text, indom and semantics handling
	python API: add log import timestamp convenience mechanisms
    - Build, infrastructure and packaging updates:
	build: tweak Makepkgs as dpkg-parsechangelog(1) is a bit picky
	build: drop gfs2 PMDA from build if distro vendor opts-out

src/pmlogger/pmlogger_merge.sh
#! /bin/sh
#
# Copyright (c) 2014,2020 Red Hat.
# Copyright (c) 1995,2003 Silicon Graphics, Inc.  All Rights Reserved.
# 
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2 of the License, or (at your
# option) any later version.
# 
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
# 
# merge a group of logfiles, e.g. all those for today
#
# default case, w/out arguments uses the default pmlogger filename
# conventions for today's logs, namely `date +%Y%m%d` for both the
# input-basename and the output-name
#

# Get standard environment
. $PCP_DIR/etc/pcp.env


prog=`basename $0`
tmp=`mktemp -d "$PCP_TMPFILE_DIR/pmlogger_merge.XXXXXXXXX"` || exit 1
tmpmerge=`mktemp -d $PCP_TMPFILE_DIR/pmlogger_merge.XXXXXXXXX` || exit 1
status=0
trap "rm -rf $tmp $tmpmerge; exit \$status" 0 1 2 3 15

force=false
VERBOSE=false
SHOWME=false
EXPUNGE=""
RM=rm

_abandon()
{
    echo "$prog: These error(s) are fatal, no output archive has been created."
    status=1
    exit
}

_warning()
{
    echo "$prog: Trying to continue, although output archive may be corrupted."
    force=false
}

_usage()
{
    pmgetopt --usage --progname=$prog --config=$tmp/usage
    status=1
    exit
}

cat > $tmp/usage << EOF
# Usage: [options] [input-basename ... output-name]

Options:
  -f, --force    remove input files after creating output files
  -N, --showme   perform a dry run, showing what would be done
  -V, --verbose  increase diagnostic verbosity
  -E, --expunge  expunge metrics with metadata mismatches between archives
  --help
EOF

# option parsing
ARGS=`pmgetopt --progname=$prog --config=$tmp/usage -- "$@"`
[ $? != 0 ] && exit 1

eval set -- "$ARGS"
while [ $# -gt 0 ]
do
    case "$1"
    in

	-f)	force=true
		;;

	-N)	SHOWME=true
		RM="echo + rm"
		;;

	-V)	VERBOSE=true
		;;

	-E)	EXPUNGE="-x" # for pmlogextract -x
		;;

	--)	shift
		break
		;;

	-\?)	_usage
		;;
    esac
    shift
done

if [ $# -eq 0 ]
then
    trylist=`date +%Y%m%d`
    output=$input
elif [ $# -ge 2 ]
then
    trylist=""
    while [ $# -ge 2 ]
    do
	trylist="$trylist $1"
	shift
    done
    output="$1"
else
    _usage
fi

fail=false
mergelist=""

# handle glob expansion if command line arguments contain
# shell metacharacters and handle dupicate-breaking name
# form of the base name i.e. YYYYMMDD.HH.MM-seq# ...
# and ensure no duplicates
#
rm -f $tmp/input
echo >$tmp/input
for try in $trylist
do
    # regular YYYYMMDD.HH.MM.meta files (maybe with compression)
    #
    for xxx in $try.meta*
    do
	[ "$xxx" = "$try.meta*" ] && continue
	base=`echo "$xxx" | sed -e 's/\(.*\)\.meta.*/\1/'`
	grep "^$base\$" $tmp/input >/dev/null || echo "$base" >>$tmp/input
    done
    # duplicate-breaking YYYYMMDD.HH.MM-seq#.meta files (maybe with compression)
    #
    for xxx in $try-*.meta*
    do
	[ "$xxx" = "$try-*.meta*" ] && continue
	base=`echo "$xxx" | sed -e 's/\(.\)\.meta.*/\1/'`
	grep "^$base\$" $tmp/input >/dev/null || echo "$base" >>$tmp/input
    done
done

# get $compress_suffixes variable
eval `pmconfig -L -s compress_suffixes`

# For each input archive, need to have at least .0 and .meta and
# warn if .index is missing.
# Need to handle compressed versions of all of these.
#
rm -f $tmp/was_compressed
for input in `cat $tmp/input`
do
    empty=0
    for part in index meta 0
    do
	rm -f $tmp/found
	for suff in "" $compress_suffixes
	do
	    if [ -f "$input.$part$suff" ]
	    then
		touch $tmp/found
		[ -n "$suff" ] && touch $tmp/was_compressed
		if [ ! -s "$input.$part$suff" ]
		then
		    empty=`expr $empty + 1`
		fi
		break
	    fi
	done
	if [ ! -f $tmp/found ]
	then
	    case $part
	    in
		index)
		    echo "$prog: Warning: \"index\" file missing for archive \"$input\""
		    empty=`expr $empty + 1`
		    ;;
		meta)
		    echo "$prog: Error: \"meta\" file missing for archive \"$input\""
		    fail=true
		    ;;
		0)
		    echo "$prog: Error: \"volume 0\" file missing for archive \"$input\""
		    ls ${file}*
		    fail=true
		    ;;
	    esac
	fi
    done
    if [ $empty -eq 3 ]
    then
	echo "$prog: Warning: archive \"$input\" is empty and will be skipped"
    else
	mergelist="$mergelist $input"
    fi
done

if [ -f $output.index ]
then
    echo "$prog: Error: \"index\" file already exists for output archive \"$output\""
    fail=true
fi
if [ -f $output.meta ]
then
    echo "$prog: Error: \"meta\" file already exists for output archive \"$output\""
    fail=true
fi
if [ -f $output.0 ]
then
    echo "$prog: Error: \"volume 0\" file already exists for output archive \"$output\""
    fail=true
fi

$fail && _abandon

i=0
list=""
part=0
if [ -z "$mergelist" ]
then
    $VERBOSE && echo "No archives to be merged."
else
    $VERBOSE && echo "Input archives to be merged:"
    for input in $mergelist
    do
	if [ $i -ge 35 ]
	then
	    # this limit requires of the order of 3 x 35 input + 3 x 1
	    # output = 108 file descriptors which should be well below any
	    # shell-imposed or system-imposed limits
	    #
	    $VERBOSE && echo "		-> partial merge to $tmpmerge/$part"
	    cmd="pmlogextract $EXPUNGE $list $tmpmerge/$part"
	    if $SHOWME
	    then
		echo "+ $cmd"
	    else
		if $cmd
		then
		    :
		else
		    $VERBOSE || echo "		-> partial merge to $tmpmerge/$part"
		    echo "$prog: Directory: `pwd`"
		    echo "$prog: Failed: pmlogextract $list $tmpmerge/$part"
		    _warning
		fi
	    fi
	    list=$tmpmerge/$part
	    part=`expr $part + 1`
	    i=0
	fi
	list="$list $input"
	$VERBOSE && $PCP_ECHO_PROG $PCP_ECHO_N "	$input""$PCP_ECHO_C"
	numarch=`echo $input.[0-9]* | wc -w | sed -e 's/  *//g'`
	if [ $numarch -gt 1 ]
	then
	    $VERBOSE && echo " ($numarch volumes)"
	else
	    $VERBOSE && echo
	fi
	i=`expr $i + 1`
    done

    cmd="pmlogextract $EXPUNGE $list $output"
    if $SHOWME
    then
	echo "+ $cmd"
    else
	if $cmd
	then
	    if [ -f $tmp/was_compressed ]
	    then
		# some component of the input archive(s) was
		# compressed, so compress the output archive
		#
		$VERBOSE && echo "Compressing $output"
		if pmlogcompress $output
		then
		    :
		else
		    echo "$prog: Failed: pmlogcompress $output"
		    _warning
		fi
	    fi
	else
	    echo "$prog: Directory: `pwd`"
	    echo "$prog: Failed: pmlogextract $list $output"
	    _warning
	fi
	$VERBOSE && echo "Output archive files:"
	for file in $output.meta $output.index $output.0
	do
	    for suff in "" $compress_suffixes
	    do
		if [ -f "$file$suff" ]
		then
		    file="$file$suff"
		    break
		fi
	    done
	    if [ -f $file ]
	    then
		$VERBOSE && LC_TIME=POSIX ls -l $file
	    else
		echo "$prog: Error: file \"$file\" not created"
		force=false
	    fi
	done
    fi
fi

if $force
then
    $VERBOSE && $PCP_ECHO_PROG $PCP_ECHO_N "Removing input archive files ""$PCP_ECHO_C"
    for input in `cat $tmp/input`
    do
	if $VERBOSE
	then
	    for file in $input.index* $input.meta* $input.[0-9]*
	    do
		$PCP_ECHO_PROG $PCP_ECHO_N ".""$PCP_ECHO_C"
	    done
	fi
	eval $RM -f $input.index* $input.meta* $input.[0-9]*
    done
    $VERBOSE && echo " done"
fi

exit
